### ISSUE TITLE
[sklearn#33958] META: Use narwhals in codebase

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#370]
**Mirrored from [scikit-learn#33958](https://github.com/scikit-learn/scikit-learn/issues/33958)**

Original title: META: Use narwhals in codebase
Original labels: Meta-issue
Created: 2026-05-09T16:33:36Z

---

#31127 was merged and so we can now rely on narwhals.

#### Reasons for more narwhals (nw)
- NW simplifies internal code by treating all dataframes (pandas, polars, arrow) with a unified API
- NW can make code faster:
  - for categorical feature support in `OrdinalEncoder` and other places
  - by use of narwhals expression API instead of array operations, e.g. group_by(..).agg(..)
- Possibly (not yet decided) narwhals-in, narwhals-out (set_output for `transform`)

#### TODO
- [ ] maintainability PRs
  - +658 -764 #31127
  - +34 -33 #33971
  - +21 -10 #34051
  - +16 -54 #34089
- [ ] Add entry for "dataframe" in glossary
- [ ] Extend user guide on [data interoperability](https://scikit-learn.org/dev/data_interoperability.html#data-interoperability)
- [ ] Remove use of `iloc`
- [ ] Extend many pandas-only checks to dataframe

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [32](https://github.com/godofecht/flow-scikit/issues/32) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [33](https://github.com/godofecht/flow-scikit/issues/33) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [34](https://github.com/godofecht/flow-scikit/issues/34) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
